// @kissmybutton/motorcortex/dist/_DescriptiveIncidents/decorators/resize.js
"use strict";

var _CONST = require('../_CONST');

var _TREECONST = require('../../_Tree/_CONST');

function resize(payload) {
  payload.descriptor.value = function (newSize) {
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {
      check: true
    };

    if (newSize === this.duration) {
      return {
        result: true,
        meta: {
          unprocessed: true
        }
      };
    }

    if (newSize <= 0) {
      return {
        result: false,
        reason: 'Size must always be > 0'
      };
    }

    if (options.check === true && this.hasParent) {
      var mcCheck = this.putMessageOnPipe('checkResize', {
        id: this.id,
        newSize: newSize,
        fraction: newSize / this.duration
      }, _CONST._TARGETTYPES._DECISIONAUTHORITY, {
        selfExecute: false,
        direction: _TREECONST._DIRECTIONS._UP
      });

      if (mcCheck.response.result === false) {
        return mcCheck.response;
      }
    }

    this.setNewDuration(newSize);
    return {
      result: true
    };
  };
}

module.exports = resize;
